### DISCUSSION TITLE
next build fails with node:4268) UnhandledPromiseRejectionWarning: Error: Cannot find module 'ts-pnp'

### DISCUSSION THREAD

**Discussion by wakemup** [vercel/next.js#12266] [Help]
I have projects built with next that have been dormant for a month or two.  I went to build them today and the next build fails with:

(node:4268) UnhandledPromiseRejectionWarning: Error: Cannot find module 'ts-pnp'

## To Reproduce

I had the same results in a completely bare bones project with only one single file - a package.json file and the following contents:

{
  "scripts": {
    "build": "next build"
  },
  "dependencies": {
    "express": "^4.17.1",
    "react": "^16.12.0",
    "react-dom": "^16.12.0",
    "next": "^9.3.5"
  }
}

If I run npm run build, I get the following output:

> next build

(node:15092) UnhandledPromiseRejectionWarning: Error: Cannot find module 'ts-pnp'
Require stack:
- C:\TestNext\node_modules\pnp-webpack-plugin\index.js
- C:\TestNext\node_modules\next\dist\build\webpack-config.js
- C:\TestNext\node_modules\next\dist\build\index.js
- C:\TestNext\node_modules\next\dist\cli\next-build.js
- C:\TestNext\node_modules\next\dist\bin\next
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:794:15)
    at Function.Module._load (internal/modules/cjs/loader.js:687:27)
    at Module.require (internal/modules/cjs/loader.js:849:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (C:\TestNext\node_modules\pnp-webpack-plugin\index.js:2:29)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Module.require (internal/modules/cjs/loader.js:849:19)
(node:15092) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:15092) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

## Expected behavior

A successful build

## System information

Windows 10
Visual Studio Code
npm version 6.12.0,
Next.js v 9.3,
Node.js v12.13

## Additional context

I sure would appreciate any insight anyone can give me.  I'm in a bit of a pickle here!

I'm not sure why this is suddenly happening and Google turns up nothing.  Sure seems like it's got something to do with Typescript, as it's the ts-pnp.  It's also pointing at pnp-webpack-plugin.  I'll be honest and say I don't really know what either of those are - or how webpack works.  I have been successfully experimenting with Typescript in non-Next projects and installed Nodemon, etc.  But the only global packages I have installed, from what I can tell are from this command:

npm list -g -depth=0

+-- eslint@6.8.0
+-- nodemon@2.0.2
+-- semistandard@14.2.0
`-- standard@14.3.1

**Comment by timneutkens** [comment]
How did you install the dependencies? and how are you running the build?

**Reply by wakemup** [reply]
![image](https://user-images.githubusercontent.com/54385834/80490019-5b323100-8926-11ea-8eef-1d617ec13467.png)

**Reply by wakemup** [reply]
I disabled all my VSCode extensions and tried again.  Same result.  **So then I tried changing the Next version to see if I could get it to go away.**  If I change the Next version in package.json from "^9.3.6" to "8.1.0", the "npm run build" command (which calls "next build") succeeds.

I am starting to think that whatever is putting these lines below into my package-lock.json file when I run "npm install" from the terminal is what is causing me issues:

"pnp-webpack-plugin": {
      "version": "1.5.0",
      "resolved": "https://registry.npmjs.org/pnp-webpack-plugin/-/pnp-webpack-plugin-1.5.0.tgz",
      "integrity": "sha512-jd9olUr9D7do+RN8Wspzhpxhgp1n6Vd0NtQ4SFkmIACZoEL1nkyAdW9Ygrinjec0vgDcWjscFQQ1gDW8rsfKTg==",
      "requires": {
        "ts-pnp": "^1.1.2"
      }
    },

These lines are NOT present if I used next version "8.1.0", but if I change to latest or 9.1.0 or ^9.3.6, they are there.  I haven't zeroed in on the exact version, however.

I'm wondering if you or anyone else is seeing this behavior - are you able to reproduce using my do-nothing setup?

FInally, I tried adding ts-pnp to the dependency list:      "ts-pnp": "^1.1.2"

After deleting the node_modules directory, running "npm install" and "npm run build", same error:

> next build

(node:23148) UnhandledPromiseRejectionWarning: Error: Cannot find module 'ts-pnp'
Require stack:
..........
...........

**Reply by wakemup** [reply]
Ok, well, I completely punted and uninstalled all my Visual Studio Code extensions, uninstalled Visual Studio Code itself, deleted all remnants of both as far as I could figure.  Deleted all global packages installed as far as I could figure.  Then I re-installed VS Code and tried again.  This time it succeeded.  But I do get about 2 screens full of warnings on the npm install.

I am pretty much a newbie to open source.  So I don't understand all the ins-and-outs here -- and I'm sure it's probably at least largely my fault.  But I am no newbie software developer.  I have been a C++, SQL & C# developer in an R & D wing of a Fortune 50 company for 25 years.  Presumably you folks want people to use NextJS.  And to be sure, I find it interesting.  I was pointed to a GitHub project using Next and based a demo I made for a client on that project.  But it makes me look pretty bad when I can't figure out how to build the project after it has sat for 6 weeks.

It's quite likely that it is something in my environment, no question.  And my drastic actions above bear this out.  But what I don't understand is why I can't get any help.  I also don't like that I get 2 screens of errors on a completely bare bones project install - on a seemingly clean environment.  This seems very odd to me.  Maybe I'm the naive one and that's how open source works.  But if it is, maybe it's not for me and maybe I need to return to my comfort zone or build from scratch without Next or any other package.

I'm not trying to be snippy - although I'm sure it comes across that way.  I'm frustrated.  I am trying to use Next, but it seems that you have to be someone that can spelunk arcane code or a code contributor to get help.  If I wanted to do that, I would write my own thing.  I just want to drop in a framework and have it work.  Is there someplace maybe you could point me to help me understand if I'm on crank here and my environment is so screwed up and causing these warnings?  Or a tutorial on what the heck would be causing things to get injected into my package-lock.json file in the pre-nuked environment?  I'm not expecting hand-holding.  But I guess I was hoping for some direction.

You guys do great work and I'm sure are rockstar software engineers.  But if you want guys like me to use Next, you'll want to consider some newbie (to Node/Next/open-source) support.
